src/app/application/services/data-source/SharedPool.ts

```diff
@@ -0,0 +1,45 @@
+type Init<K, T> = (key: K) => T | undefined
+type Del<K, T> = (key: K, item: T) => void
+export default class SharedPool<K, T> {
+  pool: Map<K, T> = new Map()
+  usage: Map<T, symbol[]> = new Map()
```

Comment:
Should this be a set of symbol?

Reply:
Try doing: `Map<K, CountingReference[T, symbol]>`

Reply:
I kept the two separate `pool`, `usage` as I 'think' we still need that, but I changed things to use a Set rather than just an array 👍 , I'm just checking with you offline to see if we definitely need the two separate `pool`, `usage`

Reply:
I think we settled on this one as it is now, lemme know if not